source/common/websocket/codec.cc

```diff
@@ -0,0 +1,220 @@
+  } else if (num_remaining_masking_key_bytes_ > 0) {
+    state_ = State::FrameHeaderMaskingKey;
+  } else {
+    frameDataStart();
+  }
+  return 1;
+}
+
+uint8_t Decoder::doDecodeExtendedLength(absl::Span<const uint8_t>& data) {
+  uint64_t bytes_to_decode = data.length() <= num_remaining_extended_length_bytes_
+                                 ? data.length()
+                                 : num_remaining_extended_length_bytes_;
+  uint8_t size_of_extended_length =
+      state_ == State::FrameHeaderExtendedLength16Bit ? kPayloadLength16Bit : kPayloadLength64Bit;
+  uint8_t shift_of_bytes = size_of_extended_length - num_remaining_extended_length_bytes_;
+  uint8_t* destination = reinterpret_cast<uint8_t*>(&length_) + shift_of_bytes;
+
+  ASSERT(shift_of_bytes >= 0);
+  ASSERT(shift_of_bytes < size_of_extended_length);
+  memcpy(destination, data.data(), bytes_to_decode); // NOLINT(safe-memcpy)
+  num_remaining_extended_length_bytes_ -= bytes_to_decode;
+
+  if (num_remaining_extended_length_bytes_ == 0) {
+    length_ = state_ == State::FrameHeaderExtendedLength16Bit ? htobe16(length_) : htobe64(length_);
+    if (num_remaining_masking_key_bytes_ > 0) {
+      state_ = State::FrameHeaderMaskingKey;
+    } else {
+      frameDataStart();
+    }
+  }
+  return bytes_to_decode;
+}
+
+uint8_t Decoder::doDecodeMaskingKey(absl::Span<const uint8_t>& data) {
+  if (!frame_.masking_key_.has_value()) {
+    frame_.masking_key_ = 0;
+  }
+  uint64_t bytes_to_decode = data.length() <= num_remaining_masking_key_bytes_
+                                 ? data.length()
+                                 : num_remaining_masking_key_bytes_;
+  uint8_t shift_of_bytes = kMaskingKeyLength - num_remaining_masking_key_bytes_;
+  uint8_t* destination =
+      reinterpret_cast<uint8_t*>(&(frame_.masking_key_.value())) + shift_of_bytes;
+  ASSERT(shift_of_bytes >= 0);
+  ASSERT(shift_of_bytes < kMaskingKeyLength);
+  memcpy(destination, data.data(), bytes_to_decode); // NOLINT(safe-memcpy)
+  num_remaining_masking_key_bytes_ -= bytes_to_decode;
+
+  if (num_remaining_masking_key_bytes_ == 0) {
+    frame_.masking_key_ = htobe32(frame_.masking_key_.value());
+    frameDataStart();
+  }
+  return bytes_to_decode;
+}
+
+uint64_t Decoder::doDecodePayload(absl::Span<const uint8_t>& data) {
+  uint64_t remain_in_buffer = data.length();
+  uint64_t bytes_decoded = 0;
+  if (remain_in_buffer <= length_) {
+    if (decode_payload_) {
+      frameData(data.data(), remain_in_buffer);
+    }
+    bytes_decoded += remain_in_buffer;
+    length_ -= remain_in_buffer;
+  } else {
+    if (decode_payload_) {
+      frameData(data.data(), length_);
+    }
+    bytes_decoded += length_;
+    length_ = 0;
+  }
+  if (length_ == 0) {
+    state_ = State::FrameFinished;
+  }
+  return bytes_decoded;
+}
+
+absl::optional<std::vector<Frame>> Decoder::decode(const Buffer::Instance& input) {
+  absl::optional<std::vector<Frame>> output = std::vector<Frame>();
+  resetDecoder();
```

Comment:
The `decodeData` of HTTP filter will be called if there is a read event for body. Will it possible that a frame is split and multiple `decodeData` are called for single frame? Then this resetDecoder seems like will break something?